GeoLib/SurfaceGrid.cpp

```diff
@@ -0,0 +1,207 @@
+/*
+ * \date 2012-09-22
+ * \brief Definition of the SurfaceGrid class.
+ *
+ * \copyright
+ * Copyright (c) 2012-2015, OpenGeoSys Community (http://www.opengeosys.org)
+ *            Distributed under a Modified BSD License.
+ *              See accompanying file LICENSE.txt or
+ */
+
+#include <algorithm>
+#include <bitset>
+
+#include <logog/include/logog.hpp>
+
+#include "SurfaceGrid.h"
+
+#include "Surface.h"
+#include "Triangle.h"
+
+#include "MathLib/Point3d.h"
+
+namespace GeoLib {
+
+SurfaceGrid::SurfaceGrid(Surface const*const sfc) :
+	AABB(sfc->getAABB()), _n_steps({{1,1,1}})
+{
+	// enlarge the bounding, such that the points with maximal coordinates
+	// fits into the grid
+	for (std::size_t k(0); k<3; ++k) {
+		_max_pnt[k] += std::abs(_max_pnt[k]) * 1e-6;
+		if (std::abs(_max_pnt[k]) < std::numeric_limits<double>::epsilon()) {
```

Comment:
In the case both values, `_min_pnt[k]` and `_max_pnt[k]`, are small positive values the property `_min_pnt[k] <= _max_pnt[k]` holds true and the new value should never be negative.